singer_sdk/sinks/core.py

```diff
@@ -401,3 +401,12 @@ def activate_version(self, new_version: int) -> None:
             "ACTIVATE_VERSION message received but not implemented by this target. "
             "Ignoring."
         )
+
+    @abc.abstractmethod
+    def clean_up(self) -> None:
+        """Perform any clean up actions required at end of a stream.
+
+        Implementations should ensure that clean up does not affect resources
+        that may be in use from other instances of the same sink. Stream name alone
+        should not be relied on, it's recommended to use a uuid as well."""
+        pass
```

Comment:
@z3z1ma tests are complaining because existing subclasses don't implement this method:

```
TypeError: Can't instantiate abstract class SampleCSVTargetSink with abstract method clean_up
```

The base case (noop) is probably OK for most use cases, so there'd be no need to force implementations of this method. Otherwise this will introduce a breaking change for SDK-based targets in the wild. If that's what we want _eventually_, we could emit a warning for targets not implementing this method, e.g.

  ```suggestion
    def clean_up(self) -> None:
        """Perform any clean up actions required at end of a stream.

        Implementations should ensure that clean up does not affect resources
        that may be in use from other instances of the same sink. Stream name alone
        should not be relied on, it's recommended to use a uuid as well."""
        warnings.warn(
            "Future versions of the MeltanoSDK will expect custom target sinks to "
            + "implement the `clean_up` method.",
            category=FutureWarning,
        )
```

Reply:
Agree the simplicity in keeping this optional is preferred. Updated.